### PR TITLE
Instantiated Command Runners and Actions (Fixes #363)

### LINKED ISSUE #363: Make CommandRunners instantiated
Currently, `CommandRunners` are storing configuration in class members without being instantiated. This may cause problems in the future and should be changed, preferably by instantiating them.

### PR DESCRIPTION
**Pull request by tradej**
I have ported the Command Runners (and Actions, for consistency) to be instantiated when used. Parts of `DockerCommandRunner` have been cut out to `command_helpers.DockerHelper` because IMHO they make more sense there. Tests pass, and I have tried both the CLI and GUI and they seem to work. Still, I'd like some input from you guys. Thanks.

Fixes #363

### CODE REVIEW

File: devassistant/actions.py [hunk 1]
```diff
@@ -370,8 +367,8 @@ def run(cls, **kwargs):
         for pkg in pkgs:
             logger.info('Updating DAP {pkg} ...'.format(pkg=pkg))
             try:
-                updated = dapicli.install_dap(pkg, update=True, update_allpaths=kwargs['allpaths'],
-                                              force=kwargs['force'])
+                updated = dapicli.install_dap(pkg, update=True, update_allpaths=self.kwargs['allpaths'],
```
Comment (hroncok) [inline devassistant/actions.py]: Too long line

File: devassistant/actions.py [hunk 2]
```diff
@@ -398,16 +395,15 @@ class PkgListAction(Action):
                           help='List packages available from DAPI (not installed only)'),
     ]
 
-    @classmethod
-    def run(cls, **kwargs):
-        if [kwargs['installed'], kwargs['remote'], kwargs['available']].count(True) > 1:
+    def run(self):
+        if [self.kwargs['installed'], self.kwargs['remote'], self.kwargs['available']].count(True) > 1:
```
Comment (hroncok) [inline devassistant/actions.py]: Too long line

File: devassistant/actions.py [hunk 3]
```diff
@@ -398,16 +395,15 @@ class PkgListAction(Action):
...
             logger.error('Only one of --installed, --remote or --available '
                          'can be used simultaneously')
             return
-        if kwargs['remote'] or kwargs['available']:
-            dapicli.print_daps(simple=kwargs['simple'], skip_installed=kwargs['available'])
+        if self.kwargs['remote'] or self.kwargs['available']:
+            dapicli.print_daps(simple=self.kwargs['simple'], skip_installed=self.kwargs['available'])
```
Comment (hroncok) [inline devassistant/actions.py]: Too long line

File: devassistant/actions.py [hunk 4]
```diff
@@ -463,28 +458,27 @@ class PkgInfoAction(Action):
...
                         'This DAP is not valid, info can\'t be displayed.')
             finally:
                 logger.setLevel(old_level)
-            dapicli.print_local_dap(d, full=kwargs.get('full', False))
-        elif kwargs.get('installed'):
+            dapicli.print_local_dap(d, full=self.kwargs.get('full', False))
+        elif self.kwargs.get('installed'):
             try:
-                dapicli.print_installed_dap(kwargs['package'], full=kwargs.get('full', False))
+                dapicli.print_installed_dap(self.kwargs['package'], full=self.kwargs.get('full', False))
```
Comment (hroncok) [inline devassistant/actions.py]: Long line

File: devassistant/actions.py [hunk 5]
```diff
@@ -463,28 +458,27 @@ class PkgInfoAction(Action):
...
             except exceptions.DapiError as e:
                 logger.error(utils.exc_as_decoded_string(e))
                 raise exceptions.ExecutionException(utils.exc_as_decoded_string(e))
         else:
             try:
-                dapicli.print_dap_from_dapi(kwargs['package'], full=kwargs.get('full', False))
+                dapicli.print_dap_from_dapi(self.kwargs['package'], full=self.kwargs.get('full', False))
```
Comment (hroncok) [inline devassistant/actions.py]: Long line

File: devassistant/actions.py [hunk 6]
```diff
@@ -561,67 +553,64 @@ class AutoCompleteAction(Action):
...
-    _special_tokens = ['--debug']
-
-    @classmethod
-    def run(cls, **kwargs):
-        path = kwargs.get('path', '').split()
-
-        flags = cls._get_flags_for_path(path)
+    def __init__(self, **kwargs):
+        self.kwargs = kwargs
+        self._assistant_names = ['create', 'tweak', 'prepare', 'extra']
```
Comment (hroncok) [inline devassistant/actions.py]: Things in here, do we need to distinguish them per instance?

(also for some of the next lines)

File: devassistant/command_runners.py [hunk 7]
```diff
@@ -75,12 +76,19 @@ def matches(cls, c):
...
-        """Runs the given command.
+    def __init__(self, c):
+        """Initialize with a command c.
+
+        This method should be reimplemented in case that the command runner
+        stores configuration
 
         Args:
-            c - command to run, instance of devassistant.command.Command
+            c - command to check, instance of devassistant.command.Command
```
Comment (hroncok) [inline devassistant/command_runners.py]: command to check?

File: devassistant/command_runners.py [hunk 8]
```diff
@@ -303,43 +316,43 @@ def run(cls, c):
...
-            msg = 'Dependencies for installation must be list, got {v}.'.format(v=c.input_res)
+    def run(self):
+        if not isinstance(self.c.input_res, list):
+            msg = 'Dependencies for installation must be list, got {v}.'.format(v=self.c.input_res)
             raise exceptions.CommandException(msg)
 
         di = DependencyInstaller()
-        di.install(c.input_res, c.kwargs['__ui__'], debug=c.kwargs.get('da_debug', False))
-        return (True, c.input_res)
+        di.install(self.c.input_res, self.c.kwargs['__ui__'], debug=self.c.kwargs.get('da_debug', False))
```
Comment (hroncok) [inline devassistant/command_runners.py]: Long line

File: devassistant/command_runners.py [hunk 9]
```diff
@@ -456,23 +469,25 @@ def _dot_devassistant_write(cls, comm):
...
                            'create_repo': ['login', 'reponame', 'private'],
                            'create_and_push': ['login', 'reponame', 'private'],
                            'create_fork': ['login', 'repo_url'],
                            'push': []}
 
+    def __init__(self, c):
+        self.c = c
+        self._user = None
+        try:
+            self._gh_module = utils.import_module('github')
```
Comment (hroncok) [inline devassistant/command_runners.py]: Is it possible, that some instances will import the module and some will fail? IS there any logical reason to keep this per instance?

File: devassistant/command_runners.py [hunk 10]
```diff
@@ -929,29 +936,29 @@ def _strip_dir_prefix(cls, prefix, path):
...
 class AsUserCommandRunner(CommandRunner):
+
     @classmethod
     def matches(cls, c):
         return c.comm_type.startswith('as ')
 
     @classmethod
-    def get_user_from_comm_type(cls, comm_type):
-        split_type = comm_type.split()
+    def get_user_from_comm_type(cls, c):
```
Comment (hroncok) [inline devassistant/command_runners.py]: This method is named `..._from_comm_type` and it accepted comm_type before, now it accepts the command and it gets the type out of it, I'd suggest either keep original behavior, or rename the method to `get_user_from_command`

**Comment by hroncok** [comment]
* In general the change as whole makes sense.
 * By adding `self.`, sometimes the line gets too long (> 99 characters), I've reported some of those in-line but I stooped at some point (there were too many)
 * +1 for DockerHelper
 * Sometimes I feel that some of the class attributes/properties should remain "static" and there is no reason to save them per instance (I've commented on those in-line)
 * One thing about get_user_from_comm_type commented in-line
 * It would be good to make appropriate changes [in the documentation](http://docs.devassistant.org/en/latest/developer_documentation/devassistant_core.html#command-runners) if not in the same commit than at least in this PR

**Comment by hroncok** [comment]
I'm also not sure if this works.

```
$ da crt python lib -n foo
DevAssistant traceback (most recent call last):
Error: No DevAssistant frames to print.

Traceback (most recent call last):
  File ".../devassistant/venv/bin/da", line 9, in <module>
    load_entry_point('devassistant==0.10.1', 'console_scripts', 'da')()
  File "...devassistant/devassistant/cli/cli_runner.py", line 73, in run
    to_run(**parsed_args_decoded).run()
TypeError: 'PathRunner' object is not callable
```

**Comment by hroncok** [comment]
Apparently, the change you've made in cli/cli_runner.py is only working for actions, not for assistants.

**Comment by hroncok** [comment]
This also makes me think we need much more integration tests, including ones actually running some assistants.

**Comment by tradej** [comment]
Thanks for the comments, I'll factor them in.

**Pull request merged**